stacker/providers/aws/default.py

```diff
@@ -448,17 +452,34 @@ def generate_stack_policy_args(stack_policy=None):
 
 
 class ProviderBuilder(object):
-    """Implements a ProviderBuilder for the AWS provider."""
+    """Implements a Memoized ProviderBuilder for the AWS provider."""
 
     def __init__(self, region=None, **kwargs):
         self.region = region
         self.kwargs = kwargs
+        self.providers = {}
+        self.lock = Lock()
 
     def build(self, region=None, profile=None):
-        if not region:
-            region = self.region
-        session = get_session(region=region, profile=profile)
-        return Provider(session, region=region, **self.kwargs)
+        """Get or create the provider for the given region and profile."""
+
+        with self.lock:
+            # memoization lookup key derived by taking md5 of region + profile.
+            key = md5("{}{}".format(region, profile)).hexdigest()
+            try:
+                # assume provider is in provider dictionary.
+                return self.providers[key]
+            except KeyError:
```

Comment:
Using an exception for this seems less readable, and introduces two return paths instead of one. Can we just do:

```python
if key not in self.providers:
  self.providers[key] = ...
return self.providers[key]
```

Reply:
This is what I originally had but @phobologic suggested using the try/except block because it would be optimized for the typical case.